okhttp/src/main/java/okhttp3/internal/http2/Hpack.java

```diff
@@ -511,6 +521,37 @@ void writeInt(int value, int prefixMask, int bits) {
       out.writeByte(value);
     }
 
+    int getHeaderIndex(Header header) {
+      int dynamicHeaderIndex = Util.indexOf(dynamicTable, header);
+      if (dynamicHeaderIndex != -1) return normalizeDynamicTableIndex(dynamicHeaderIndex);
+
+      // Only search a subset of the static header table. Most entries have an empty value, so it's
+      // unnecessary to waste cycles looking at them.
+      for (int i = 1; i < 7; i++) {
+        if (equal(STATIC_HEADER_TABLE[i], header)) return i + 1;
+      }
+
+      return -1;
+    }
+
+    /** Converts a raw array index in the dynamic table array into the HPACK index address space. */
+    int normalizeDynamicTableIndex(int arrayIndex) {
+      return arrayIndex - nextHeaderIndex + STATIC_HEADER_TABLE.length;
+    }
+
+    int getHeaderNameIndex(ByteString name) {
+      Integer staticIndex = NAME_TO_FIRST_INDEX.get(name);
+      if (staticIndex != null) return staticIndex + 1;
+
+      for (int i = 0, size = dynamicTable.length; i < size; i++) {
```

Comment:
Also worth considering, given an infinite amount of time...

Right now we loop over the static & dynamic tables up to two times:
 - look for an exact match
 - look for a name match

What would happen if we looped 1x, and returned the ”best match” – the exact index if it’s found, or the header index otherwise. Figuring out a way to return that unambiguously without allocation is painful (negative vs. positive?) and might make it not worth the effort.